subprojects/functional/src/main/java/org/gradle/internal/Deferrable.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.internal;
+
+import java.util.Optional;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * An invocation to be executed at most once, but one that can be deferred.
+ *
+ * @param <T> The type which will be computed.
+ */
+public interface Deferrable<T> {
```

Comment:
Javadoc should specify the contract regarding the completion with null values, and how that works with `getCompleted` that returns an `Optional`, etc.

Reply:
We rely on the `@NonNullApi` annotation on the package for documentation. With that in mind the behavior wrt to `Optional` should be self-explanatory as well.